README.md

```diff
@@ -0,0 +1,418 @@
+          "source": {
+            "api_key": "fake:api-key",
+            "hosts": [
+              "https://localhost:9092"
+            ],
+            "type": "elasticsearch"
+          },
+          "type": "elasticsearch"
+        },
+        "logLevel": "INFO"
+      }
+    ],
+    "features": {
+      "source": {
+        "agent": {
+          "features": {
+            "fqdn": {
+              "enabled": false
+            }
+          }
+        }
+      },
+      "fqdn": {}
+    },
+    "featuresIdx": "2",
+    "component": {
+      "limits": {
+        "source": {
+          "go_max_procs": 0
+        }
+      }
+    },
+    "componentIdx": "2"
+  }
+}
+```
+
+</details>
+
+3. The `logfile` component checks in again reporting that it has succesfully applied the latest configuration.
+
+```json
+{
+  "@timestamp": "2023-08-15T19:41:41.009Z",
+  "observed": {
+    "token": "27e63445-dff2-48a7-927d-3113ab50769b",
+    "units": [
+      {
+        "id": "log-default-logfile-system-8ec9254c-0f1e-4166-965b-549c2818a0b0",
+        "configStateIdx": "1",
+        "message": "Starting"
+      },
+      {
+        "id": "log-default",
+        "type": "OUTPUT",
+        "configStateIdx": "1",
+        "state": "HEALTHY",
+        "message": "Healthy"
+      }
+    ],
+    "featuresIdx": "2",
+    "componentIdx": "2"
+  }
+}
+```
+
+4. The agent responds with an abbreviated `CheckinExpected` message indicating the component is up to date.
+
+```json
+{
+  "@timestamp": "2023-08-15T19:41:41.009Z",
+  "expected": {
+    "units": [
+      {
+        "id": "log-default-logfile-system-8ec9254c-0f1e-4166-965b-549c2818a0b0",
+        "state": "HEALTHY",
+        "configStateIdx": "1",
+        "logLevel": "INFO"
+      },
+      {
+        "id": "log-default",
+        "type": "OUTPUT",
+        "state": "HEALTHY",
+        "configStateIdx": "1",
+        "logLevel": "INFO"
+      }
+    ],
+    "features": {
+      "source": {
+        "agent": {
+          "features": {
+            "fqdn": {
+              "enabled": false
+            }
+          }
+        }
+      },
+      "fqdn": {}
+    },
+    "featuresIdx": "2",
+    "component": {
+      "limits": {
+        "source": {
+          "go_max_procs": 0
+        }
+      }
+    },
+    "componentIdx": "2"
+  },
+}
+```
+
+### Sample Logging Code
+
+To log control protocol messages in the format above in Go, code like the following can be used with
+the [zap](https://github.com/uber-go/zap) logger. 
+
+```go
+    // When logging either the agent or client side of the exchange both an observed and expected response
+    // are likely be available at the same time.
+    var expected *proto.CheckinExpected,
+    var observed *proto.CheckinObserved,
+
+    obsJSON, _ := protojson.Marshal(observed)
+    expJSON, _ := protojson.Marshal(expected)
+    c.logger.Infow("Checkin",
+        zap.Any("observed", json.RawMessage(obsJSON)),
+        zap.Any("expected", json.RawMessage(expJSON)),
+    )
+```
```

Comment:
Would it be possible to include this code in Agents built with `DEV=true`?

Reply:
If we hadn't accidentally released the Beats build with DEV=true previously I'd say yes, but we did do this in the early 8.x releases. Doing that with these would leak credentials in the agent logs unfortunately.

Reply:
I'll see if I can think of an easy way to turn this on and off.